src/librustc_data_structures/transitive_relation.rs

```diff
@@ -0,0 +1,566 @@
+        // (a, b) are in a consistent relative order
+        if a > b {
+            mem::swap(&mut a, &mut b);
+        }
+
+        let lub_indices = self.take_closure(|closure| {
+            // Easy case is when either a < b or b < a:
+            if closure.contains(a.0, b.0) {
+                return vec![b.0];
+            }
+            if closure.contains(b.0, a.0) {
+                return vec![a.0];
+            }
+
+            // Otherwise, the tricky part is that there may be some c
+            // where a < c and b < c. In fact, there may be many such
+            // values. So here is what we do:
+            //
+            // 1. Find the vector `[X | a < X && b < X]` of all values
+            //    `X` where `a < X` and `b < X`.  In terms of the
+            //    graph, this means all values reachable from both `a`
+            //    and `b`. Note that this vector is also a set, but we
+            //    use the term vector because the order matters
+            //    to the steps below.
+            //    - This vector contains upper bounds, but they are
+            //      not minimal upper bounds. So you may have e.g.
+            //      `[a, b, tcx, x]` where `a < tcx` and `b < tcx` and
+            //      `x < a` and `x < b`. This could be reduced to
+            //      just `[x]`.
+            // 2. Pare down the vector using `pare_down`. This will
+            //    remove elements from the vector that can be reached
+            //    by an earlier element.
+            //    - In the example above, this would convert
+            //      `[a, b, tcx, x]` to `[a, b, x]`. Note that `x`
+            //      remains because `x < a` but not `a < x.`
+            // 3. Reverse the vector and repeat the pare down process.
+            //    - In the example above, we would reverse to
+            //      `[x, b, a]` and then pare down to `[x]`.
+            // 4. Reverse once more just so that we yield a vector in
+            //    increasing order of index. Maybe this is silly.
+            //
+            // I believe this algorithm yields a minimal set. The
+            // argument is that, after step 2, we know that no element
+            // can reach its successors (in the vector, not the graph).
+            // After step 3, we know that no element can reach any of
+            // its predecesssors (because of step 2) nor successors
+            // (because we just called `pare_down`)
+
+            let mut candidates = closure.intersection(a.0, b.0); // (1)
+            pare_down(&mut candidates, closure); // (2)
+            candidates.reverse(); // (3a)
+            pare_down(&mut candidates, closure); // (3b)
+            candidates.reverse(); // (4)
+            candidates
+        });
+
+        lub_indices.into_iter()
+                   .map(|i| &self.elements[i])
+                   .collect()
+    }
+
+    fn take_closure<OP,R>(&self, op: OP) -> R
```

Comment:
I'd prefer the name `with_closure` over `take_closure`, since the closure in question is not being taken out of the cell permanently, but rather temporarily.

(But maybe the name `take` remains good due to the use of `borrow_mut` internally, and thus an attempt to actually use this implementation from multiple threads will hit a panic there... not that I expect us to be doing that any time soon.)